Currently only supports spark2.3.4?

Currently only supports spark2.3.4, spark2.4.5 does not support， can you help me?

: java.lang.NoClassDefFoundError: org/apache/spark/sql/sources/v2/reader/DataReaderFactory
	at bi.deep.DruidSource.createReader(DruidSource.scala:12)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$SourceHelpers.createReader(DataSourceV2Relation.scala:155)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$.create(DataSourceV2Relation.scala:172)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:204)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:167)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: org.apache.spark.sql.sources.v2.reader.DataReaderFactory
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 16 more

--- a/build.sbt
+++ b/build.sbt
@@ -4,7 +4,7 @@ ThisBuild / version := "0.2.0-SNAPSHOT"
 
 ThisBuild / scalaVersion := "2.11.12"
 
-val sparkVersion = "2.3.4"
+val sparkVersion = "2.4.8"
 val hadoopVersion = "2.7.2"
 val druidVersion = "0.22.1"
 

--- a/src/main/scala/bi/deep/DruidDataReader.scala
+++ b/src/main/scala/bi/deep/DruidDataReader.scala
@@ -1,17 +1,17 @@
 package bi.deep
 
 import org.apache.druid.segment.{DruidRowConverter, QueryableIndexIndexableAdapter}
-import org.apache.spark.sql.Row
-import org.apache.spark.sql.sources.v2.reader.DataReader
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.sources.v2.reader.InputPartitionReader
 import org.apache.spark.sql.types.{StructField, StructType}
 
 import java.io.File
 
 
 case class DruidDataReader(filePath: String, schema: StructType, config: Config)
-  extends DataReader[Row] with DruidSegmentReader {
+  extends InputPartitionReader[InternalRow] with DruidSegmentReader {
 
-  private var current: Option[Row] = None
+  private var current: Option[InternalRow] = None
   private val targetRowSize: Int = schema.size
   private val timestampIdx: Option[Int] = {
     if (config.druidTimestamp != "") Option(schema.fieldIndex(config.druidTimestamp))
@@ -28,7 +28,7 @@ case class DruidDataReader(filePath: String, schema: StructType, config: Config)
       .zipWithIndex
       .filter(f => rowFieldsNames.contains(f._1.name))
 
-    DruidRowConverter(qiia, segmentSchema, targetRowSize, filteredTargetFields, timestampIdx)
+    DruidRowConverter(qiia, schema, segmentSchema, targetRowSize, filteredTargetFields, timestampIdx)
   }
 
   override def next(): Boolean = {
@@ -37,7 +37,7 @@ case class DruidDataReader(filePath: String, schema: StructType, config: Config)
     hadNext
   }
 
-  override def get(): Row = current.get
+  override def get(): InternalRow = current.get
 
   override def close(): Unit = rowConverter.close()
 }

--- a/src/main/scala/bi/deep/DruidDataReaderFactory.scala
+++ b/src/main/scala/bi/deep/DruidDataReaderFactory.scala
@@ -1,12 +1,11 @@
 package bi.deep
 
-import org.apache.spark.sql.Row
-import org.apache.spark.sql.sources.v2.reader.{DataReader, DataReaderFactory}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.sources.v2.reader.{InputPartition, InputPartitionReader}
 import org.apache.spark.sql.types.StructType
 
 
 case class DruidDataReaderFactory(filePath: String, schema: StructType, config: Config)
-  extends DataReaderFactory[Row] {
-
-  override def createDataReader(): DataReader[Row] = DruidDataReader(filePath, schema, config)
+  extends InputPartition[InternalRow] {
+  override def createPartitionReader(): InputPartitionReader[InternalRow] = DruidDataReader(filePath, schema, config)
 }

--- a/src/main/scala/bi/deep/DruidDataSourceReader.scala
+++ b/src/main/scala/bi/deep/DruidDataSourceReader.scala
@@ -1,8 +1,10 @@
 package bi.deep
 
+import org.apache.druid.common.guava.GuavaUtils
 import org.apache.hadoop.fs.LocatedFileStatus
 import org.apache.spark.sql.Row
-import org.apache.spark.sql.sources.v2.reader.{DataReaderFactory, DataSourceReader}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.sources.v2.reader.{DataSourceReader, InputPartition}
 import org.apache.spark.sql.types.StructType
 
 import java.util
@@ -16,16 +18,17 @@ class DruidDataSourceReader(config: Config) extends DataSourceReader {
   private lazy val schema: StructType = readSchema()
 
 
+
   override def readSchema(): StructType = {
     val schemaReader = new DruidSchemaReader(mainPath, config)
     schemaReader.calculateSchema(filesPaths)
   }
 
-  private def fileToReaderFactory(file: LocatedFileStatus): DataReaderFactory[Row] = {
+  private def fileToReaderFactory(file: LocatedFileStatus): InputPartition[InternalRow] = {
     DruidDataReaderFactory(file.getPath.toString, schema, config)
   }
 
-  override def createDataReaderFactories(): util.List[DataReaderFactory[Row]] = {
+  override def planInputPartitions(): util.List[InputPartition[InternalRow]] = {
     filesPaths.toList.map(fileToReaderFactory).asJava
   }
 }

--- a/src/main/scala/bi/deep/DruidSegmentReader.scala
+++ b/src/main/scala/bi/deep/DruidSegmentReader.scala
@@ -1,13 +1,21 @@
 package bi.deep
 
 import bi.deep.Config.DRUID_SEGMENT_FILE
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.google.inject.Key
 import org.apache.commons.codec.digest.DigestUtils.sha1Hex
 import org.apache.commons.io.FileUtils
 import org.apache.druid.common.config.NullHandling
+import org.apache.druid.guice.annotations.Json
+import org.apache.druid.guice.{GuiceInjectableValues, GuiceInjectors}
 import org.apache.druid.jackson.DefaultObjectMapper
+import org.apache.druid.java.util.emitter.EmittingLogger
 import org.apache.druid.query.DruidProcessingConfig
 import org.apache.druid.segment.IndexIO
+import org.apache.druid.segment.column.ColumnConfig
 import org.apache.hadoop.fs.Path
+import org.apache.log4j.{Level, Logger}
+import org.slf4j.LoggerFactory
 
 import java.io.File
 import java.nio.file.Files
@@ -18,8 +26,18 @@ trait DruidSegmentReader {
   DruidSegmentReader.init()
 
   def indexIO: IndexIO = {
-    new IndexIO(
-      new DefaultObjectMapper(),
+    // https://www.tabnine.com/code/java/classes/org.apache.druid.guice.GuiceInjectors
+    val injector = GuiceInjectors.makeStartupInjector()
+    val injectables = new GuiceInjectableValues(injector)
+
+    // Mark as ERROR level to not display warnings with unreadable segment
+    Logger.getLogger(classOf[IndexIO]).setLevel(Level.ERROR)
+
+    val mapper = injector
+      .getInstance(Key.get(classOf[ObjectMapper], classOf[Json]))
+      .setInjectableValues(injectables)
+
+    new IndexIO(mapper,
       new DruidProcessingConfig {
         override def getFormatString: String = "processing-%s"
       }

--- a/src/main/scala/bi/deep/LatestSegmentSelector.scala
+++ b/src/main/scala/bi/deep/LatestSegmentSelector.scala
@@ -50,8 +50,8 @@ object LatestSegmentSelector {
 
   def listPrimaryPartitions(dates: List[(LocalDate, LocalDate)], path: String): List[String] = {
     path.split(':')(0) match {
-      case "s3a" | "s3" => dates.map(x => path + s"${x._1}T00:00:00.000Z_${x._2}T00:00:00.000Z")
       case "hdfs" => dates.map(x => path + s"${x._1}T00_00_00.000Z_${x._2}T00_00_00.000Z")
+      case _ => dates.map(x => path + s"${x._1}T00:00:00.000Z_${x._2}T00:00:00.000Z")
     }
   }
 }

--- a/src/main/scala/org/apache/druid/segment/DruidRowConverter.scala
+++ b/src/main/scala/org/apache/druid/segment/DruidRowConverter.scala
@@ -1,54 +1,63 @@
 package org.apache.druid.segment
 
-import org.apache.spark.sql.Row
-import org.apache.spark.sql.catalyst.expressions.GenericRow
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{SpecificInternalRow, UnsafeProjection}
+import org.apache.spark.sql.catalyst.util.ArrayData
 import org.apache.spark.sql.types.{ArrayType, StructField, StructType}
+import org.apache.spark.unsafe.types.UTF8String
 
 import java.util
+import scala.collection.JavaConverters.asScalaBufferConverter
 
-class DruidRowConverter(rowIt: QueryableIndexIndexableAdapter#RowIteratorImpl, segmentSchema: StructType,
+class DruidRowConverter(rowIt: QueryableIndexIndexableAdapter#RowIteratorImpl, fullSchema: StructType, segmentSchema: StructType,
                         targetRowSize: Int, filteredTargetFields: Array[(StructField, Int)], timestampIdx: Option[Int])
-  extends Iterator[Row] with AutoCloseable {
+  extends Iterator[InternalRow] with AutoCloseable {
+
+  private val unsafeProjection = UnsafeProjection.create(fullSchema)
 
   override def hasNext: Boolean = rowIt.moveToNext
 
-  override def next(): Row = druidRowToSparkRow(rowIt.getPointer)
+  override def next(): InternalRow = druidRowToSparkRow(rowIt.getPointer)
 
   override def close(): Unit = rowIt.close()
 
-  private def druidRowToSparkRow(row: RowPointer): Row = {
-    val targetRow = new Array[Any](targetRowSize)
+  private def druidRowToSparkRow(row: RowPointer): InternalRow = {
+    val internal = new SpecificInternalRow(fullSchema)
 
     filteredTargetFields.foreach { case (targetField, targetFieldIdx) =>
       val rowFieldIndex = segmentSchema.fieldIndex(targetField.name)
-      val value = row.getDimensionSelector(rowFieldIndex).getObject
-      targetRow(targetFieldIdx) = targetField.dataType match {
+      val selector = row.getDimensionSelector(rowFieldIndex)
+      val value = selector.getObject
+
+      internal(targetFieldIdx) = targetField.dataType match {
         case _: ArrayType => value match {
           case null => null
-          case arrValue: util.List[_] => arrValue.toArray
-          case strValue: String => Array(strValue)
+          case arrValue: util.List[String] => ArrayData.toArrayData(arrValue.asScala.map(UTF8String.fromString).toArray)
+          case strValue: String => ArrayData.toArrayData(Array(UTF8String.fromString(strValue)))
           case _ =>
             throw new ClassCastException(s"Invalid type: ${value.getClass.toString}")
         }
         case _ => value match {
           case arrValue: util.List[_] =>
             throw new RuntimeException(s"Single value expected, but found ${arrValue.size} values: ${arrValue.toString}")
-          case strValue => strValue
+          case strValue: String => UTF8String.fromString(strValue)
+          case null => null
         }
       }
     }
     if (timestampIdx.isDefined)
-      targetRow(timestampIdx.get) = row.getTimestamp
+      internal(timestampIdx.get) = row.getTimestamp
 
-    new GenericRow(targetRow)
+    unsafeProjection(internal)
+    //    internal
   }
 }
 
 
 object DruidRowConverter {
 
-  def apply(qiia: QueryableIndexIndexableAdapter, segmentSchema: StructType, targetRowSize: Int,
+  def apply(qiia: QueryableIndexIndexableAdapter, fullSchema: StructType, segmentSchema: StructType, targetRowSize: Int,
             filteredTargetFields: Array[(StructField, Int)], timestampIdx: Option[Int]): DruidRowConverter = {
-    new DruidRowConverter(qiia.getRows, segmentSchema, targetRowSize, filteredTargetFields, timestampIdx)
+    new DruidRowConverter(qiia.getRows, fullSchema, segmentSchema, targetRowSize, filteredTargetFields, timestampIdx)
   }
 }

